src/server/gc_worker/compaction_filter.rs

```diff
@@ -172,106 +179,170 @@ struct WriteCompactionFilter {
-        };
-
-        WriteCompactionFilter {
+        let mut filter = WriteCompactionFilter {
             safe_point,
-            engine,
-            write_batch,
+            handle_mvcc_deletes: context.is_bottommost_level() || too_many_delete_marks(props),
+            engine: RocksEngine::from_db(db.clone()),
+            is_invalid: false,
+            write_batch: RocksWriteBatch::with_capacity(db, DEFAULT_DELETE_BATCH_SIZE),
+
             key_prefix: vec![],
             remove_older: false,
-            write_iter,
+            write_iter: None,
             near_seek_distance: NEAR_SEEK_LIMIT,
+
             versions: 0,
             filtered: 0,
             deleted: 0,
             total_versions: 0,
             total_filtered: 0,
             total_deleted: 0,
+        };
+
+        let iter_opts = IterOptions::default();
+        match filter.engine.iterator_cf_opt(CF_WRITE, iter_opts) {
+            Ok(iter) => filter.write_iter = Some(iter),
+            Err(e) => {
+                warn!("compaction filter can't init iterator"; "err" => ?e);
+                filter.is_invalid = true;
+            }
+        }
+        filter
+    }
+
+    fn do_filter_v2(
+        &mut self,
+        _start_level: usize,
+        key: &[u8],
+        value: &[u8],
+        value_type: CompactionFilterValueType,
+    ) -> Result<CompactionFilterDecision, String> {
+        self.near_seek_distance += 1;
+        let (key_prefix, commit_ts) = split_ts(key)?;
+        if commit_ts > self.safe_point || value_type != CompactionFilterValueType::Value {
+            return Ok(CompactionFilterDecision::Keep);
+        }
+
+        self.versions += 1;
+        if self.key_prefix != key_prefix {
+            self.key_prefix.clear();
+            self.key_prefix.extend_from_slice(key_prefix);
+            self.remove_older = false;
+            self.switch_key_metrics();
+        }
+
+        let mut filtered = self.remove_older;
+        let write = parse_write(value)?;
+        if !self.remove_older {
+            match write.write_type {
+                WriteType::Rollback | WriteType::Lock => filtered = true,
+                WriteType::Put => self.remove_older = true,
+                WriteType::Delete => {
+                    self.remove_older = true;
+                    if self.handle_mvcc_deletes {
+                        filtered = true;
+                        self.handle_delete_mark(key)?;
+                    }
+                }
+            }
         }
+
+        let decision = match (filtered, self.remove_older) {
+            (true, true) => {
+                self.handle_filtered_write(write)?;
+                self.filtered += 1;
+                let prefix = Key::from_encoded_slice(key_prefix);
+                let skip_until = prefix.append_ts(0.into()).into_encoded();
+                CompactionFilterDecision::RemoveAndSkipUntil(skip_until)
+            }
+            (true, false) => CompactionFilterDecision::Remove,
```

Comment:
Why not skip key whose  `filter` is true?

Reply:
This case is for `rollback` and `lock`. `remove_older` is false so that we can't use skip.